RestaurantReservation.API/Controllers/CustomersController.cs

```diff
@@ -0,0 +1,162 @@
+﻿using AutoMapper;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.JsonPatch;
+using Microsoft.AspNetCore.Mvc;
+using RestaurantReservation.API.Models.Customers;
+using RestaurantReservation.Db.Interfaces;
+using RestaurantReservation.Db.Models.Entities;
+using System.Text.Json;
+
+namespace RestaurantReservation.API.Controllers
+{
+    [Route("api/Customers")]
+    [ApiController]
+    [Authorize]
+    public class CustomersController : ControllerBase
+    {
+        private readonly ICustomerRepository _customerRepository;
+        private readonly IMapper _mapper;
+        private const int MaxPageSize = 5;
+
+        public CustomersController(ICustomerRepository customerRepository, IMapper Mapper)
+        {
+            _customerRepository = customerRepository;
+            _mapper = Mapper;
+        }
+
+        /// <summary>
+        /// Gets all customers with pagination.
+        /// </summary>
+        /// <param name="pageNumber">The number of the page to retrieve.</param>
+        /// <param name="pageSize">The number of items per page.</param>
+        /// <returns>A list of customer DTOs.</returns>
+        [HttpGet]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public async Task<ActionResult<IEnumerable<CustomerDto>>> GetAllCustomers(int pageNumber = 1, int pageSize = 10)
+        {
+            if (pageNumber < 1 || pageSize < 1)
+                return BadRequest($"'{nameof(pageNumber)}' and '{nameof(pageSize)}' must be greater than 0.");
+
+            pageSize = Math.Min(pageSize, MaxPageSize);
+
+            var (customers, paginationMetadata) = await _customerRepository.GetAllAsync(_ => true, pageNumber, pageSize);
```

Comment:
what do you think about accessing the repo directly in controller vs adding another layer between them ?

Reply:
Good question, Adding a service layer between the controller and repository can be a good idea, especially as the application grows,it help keep the controller focused on handling HTTP request while the service layer can take care of any business logic or additional processin before interacting with the database, tthis separation make the code more modular and easier to test since we can unit test the service layer independently, Plus if we need to add more complex logic in the future having a service layer already in place makes it simpler to manage

Reply:
Good question, Adding a service layer between the controller and repository can be a good idea, especially as the application grows,it help keep the controller focused on handling HTTP request while the service layer can take care of any business logic or additional processin before interacting with the database, tthis separation make the code more modular and easier to test since we can unit test the service layer independently, Plus if we need to add more complex logic in the future having a service layer already in place makes it simpler to manage